stock_picking_mail_incorrect_qty/models/stock_picking.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2021 Coop IT Easy SCRL fs
+#   Carmen Bianca Bakker <carmen@coopiteasy.be>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class Picking(models.Model):
+    _inherit = "stock.picking"
+
+    zero_received_move_ids = fields.One2many(
+        "stock.move",
+        string="Stock moves that were not received",
+        compute="_compute_zero_received",
+    )
+    too_few_received_move_ids = fields.One2many(
+        "stock.move",
+        string="Stock moves of which too few were received",
+        compute="_compute_too_few_received",
+    )
+    too_many_received_move_ids = fields.One2many(
+        "stock.move",
+        string="Stock moves of which too many were received",
+        compute="_compute_too_many_received",
+    )
+
+    form_view_url = fields.Char(
+        string="Form View URL", compute="_compute_form_view_url"
+    )
+
+    @api.multi
+    @api.depends("move_lines")
+    def _compute_zero_received(self):
+        for pick in self:
+            pick.zero_received_move_ids = pick._filtered_moves().filtered(
+                lambda move: move.quantity_done == 0 and move.product_qty
+            )
+
+    @api.multi
+    @api.depends("move_lines")
+    def _compute_too_few_received(self):
+        for pick in self:
+            pick.too_few_received_move_ids = pick._filtered_moves().filtered(
+                lambda move: move.quantity_done < move.product_qty
+            )
+
+    @api.multi
+    @api.depends("move_lines")
+    def _compute_too_many_received(self):
+        for pick in self:
+            pick.too_many_received_move_ids = pick._filtered_moves().filtered(
+                lambda move: move.quantity_done > move.product_qty
+            )
+
+    @api.multi
+    def _compute_form_view_url(self):
+        for pick in self:
+            pick.form_view_url = (
+                "{}/web#id={}&model=stock.picking&view_type=form".format(
+                    pick.env["ir.config_parameter"].sudo().get_param("web.base.url"),
+                    pick.id,
+                )
+            )
+
+    @api.multi
+    def action_done(self):
+        self._notify_incorrect_delivery()
+
+        return super(Picking, self).action_done()
+
+    @api.multi
+    def _notify_incorrect_delivery(self):
+        """Send a notification e-mail about the incorrect delivery."""
+        for pick in self:
+            if not any(
+                (
+                    pick.zero_received_move_ids,
+                    pick.too_few_received_move_ids,
+                    pick.too_many_received_move_ids,
+                )
+            ):
+                # TODO: Handle this case.
+                continue
+
+            self.env.ref(
+                "stock_picking_mail_incorrect_qty.mail_template_incorrect_delivery"
+            ).send_mail(pick.id)
+
+        return True
+
+    @api.multi
+    def _filtered_moves(self):
```

Comment:
You could be more descriptive on what you're filtering here I think. Or maybe to it directly in the `_compute_move_discrepancy` body since it's now called once and is a one liner.